README.md
# EasyWaveSpawner
A game developer API to add waves to their games or to existing games.

## Requirements

npm installs 
-   node packages
    - [supertest]https://www.npmjs.com/package/supertest
    - [cors]https://www.npmjs.com/package/cors
    - [express]https://www.npmjs.com/package/express
    - [knex]https://www.npmjs.com/package/knex
    - [nodemon]https://www.npmjs.com/package/nodemon
    - [pg]https://www.npmjs.com/package/pg
    - [uuid]https://www.npmjs.com/package/uuid
    - [body-parser]https://www.npmjs.com/package/body-parser
    - [jest]https://www.npmjs.com/package/jest
    
## Startup
```bash
cd api
docker-compose up
```
=> open localhost:3000

## Endpoints
-   /getAllWaves => get all the existing waves
-   /getAllGames => get all existing games
-   /getWavesFromGame/:title => get all the waves from a given game , params : gametitle
-   /getwavesbydifficulty/:difficulty => get all the waves with a certain difficulty : params : easy/medium/hard/extreme
-   /deleteWaveByid/:id => delete a specific wave by id
-   /deleteGame/:id => delete a game by id with all of its linked waves
-   /changeWave/:id => change wave properties with specific id 
-   /createGame => create new game with title and name
-   /createWave/:gameTitle => create a new wave in existing game




## License
[MIT](/LICENSE)



## Contributing
I would like to the contributing.md file in the root of of this repository.



CHANGELOG.md
# V0.2.0

## Endpoints
```
-[post] /creatGame => gives back uuid of created game for usage
```
## Tests
```
- integration tests added
- end to end tests added
```

# V0.1.0

## Endpoints
```
-[GET]     /getAllWaves => get all the existing waves
-[GET]     /getAllGames => get all existing games
-[GET]     /getWavesFromGame/:title => get all the waves from a given game , params : gametitle
-[GET]     /getwavesbydifficulty/:difficulty => get all the waves with a certain difficulty : params : easy/medium/hard/extreme
-[DELETE]  /deleteWaveByid/:id => delete a specific wave by id
-[DELETE]  /deleteGame/:id => delete a game by id with all of its linked waves
-[PATCH]   /changeWave/:id => change wave properties with specific id 
-[POST]    /createGame => create new game with title and name
-[POST]    /createWave/:gameTitle => create a new wave in existing game
```

## Tests
```
- unit tests for every endpoint
```


api/src/__tests__/server.test.js
const supertest = require("supertest");
const app = require("../server.js");
const request = supertest(app);

///SERVER TEST///
describe("server test", () => {

    test("if server exists", async (done) => {

        try {
            const response = await request.get('/')
            expect(response.status).toBe(200);
            done()
        } catch (e) {
            if (e) {
                console.log(e);
            }
        }
    })
})
///GET ALL GAMES///
describe('getAllGames', () => {
    test('If i get games back', async (done) => {
        try {
            const response = await request.get('/getAllGames')
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            console.log(error);
        }
    })


})
///GET WAVES BY GAME///
describe('getWavesFromGame/:title', () => {
    test('Caps or no caps dont matter', async (done) => {
        try {
            const response = await request.get('/getWavesFromGame/Game2')
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            // console.log(error);
        }
    })
    test("gives back not found when game doesnt exist", async (done) => {

        try {
            const response = await request.get('/getWavesFromGame/game88')
            expect(response.status).toBe(404);
            done()
        } catch (e) {
            if (e) {
                console.log(e);
            }
        }
    })

})
///GET ALL WAVES///
describe('getAllWaves', () => {

    test('If I get waves back', async (done) => {
        try {
            const response = await request.get('/getAllWaves')
            console.log(response);
            expect(response.status).toBe(200);
            done()

        } catch (error) {
            // console.log(error);
        }
    })


})
///GET ALL WAVES BY DIFFICULTY///
describe('getWavesByDifficulty', () => {

    test('If I get waves back', async done => {
        try {
            const response = await request.get('/getWavesByDifficulty/hard')
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            console.log(error);
        }
    })
    test('if difficulty is hard,medium,easy or extreme', async done => {
        try {
            const response = await request.get('/getWavesByDifficulty/somethingelsethanallowed')

            expect(response.body).toBeFalsy;
            done()
        } catch (error) {
            console.log(error);
        }
    })
    test('caps or no caps dont matter', async done => {
        try {
            const response = await request.get('/getWavesByDifficulty/HaRd')
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            console.log(error);
        }
    })


})
///DELETE ENTIRE GAME///
// describe('deleteGame', () => {

//     test('if game is deleted', async (done) => {
//         try {
//             ///delete id of game3
//             const response = await request.delete('/deleteGame/:id')
//             .then(await request.get('/getWavesFromGame/Game3'))
//             done()
//         } catch (error) {
//             console.log(error);
//         }
//     })
// })



// })
///update wave difficulty
describe('updateWaves', () => {
    let id;


    test('if wavedifficulty is updated', async (done) => {
        let data = {
            "difficulty": "easy",
            "enemy_amount": 0
        }
        try {
            const waves = await request.get("/getAllWaves")
            try {
                for (const wave of waves.body) {
                    id = wave.uuid
                }
                const response = await request.patch(`/changeWave/${id}`).send(data)
                expect(response.status).toBe(204);
                done()
            } catch (error) {
                console.log(error)
            }


            done()
        } catch (error) {
            console.log(error);
        }
    })
    test('if new difficulty is "hard","easy",extreme" or medium', async (done) => {
        let data = {
            "difficulty": "55"
        }
        try {
            const waves = await request.get("/getAllWaves")
            try {
                for (const wave of waves.body) {
                    id = wave.uuid
                }
                const response = await request.patch(`/changeWave/${id}`).send(data)
                expect(response.body).toBeFalsy;
                done()
            } catch (error) {
                console.log(error)
            }



        } catch (error) {
            console.log(error);
        }
    })
    test('caps or no caps dont matter', async (done) => {
        let data = {
            "difficulty": "HaRd"
        }
        try {
            const waves = await request.get("/getAllWaves")
            try {
                for (const wave of waves.body) {
                    id = wave.uuid
                }
                const response = await request.patch(`/changeWave/${id}`).send(data)
                expect(response.status).toBe(204);
                done()
            } catch (error) {
                console.log(error)
            }



        } catch (error) {
            console.log(error);
        }
    })
})

describe('create waves/games', () => {
    test('if game is created', async (done) => {
        let data = {
            "title": "fallguys",
            "summary": "platformer where people fight to be the last man standing"
        }

        try {
            const response = await request.post(`/createGame`).send(data)
            expect(response.status).toBe(201);
            done()
        } catch (error) {
            console.log(error)
        }
    })
    test('if game has title', async (done) => {
        let data = {
            "title": "",
            "summary": "summary of the game"
        }

        try {
            const response = await request.post(`/createGame`).send(data)
            expect(response.status).toBe(500);
            done()
        } catch (error) {
            console.log(error)
        }
    })
    test('if game already exists', async (done) => {
        let data = {
            "title": "fallguys",
            "summary": "platformer where people fight to be the last man standing"
        }

        try {
            const response = await request.post(`/createGame`).send(data)
            expect(response.status).toBe(500);
            done()
        } catch (error) {
            console.log(error);
        }
    })
    test('if wave is created', async (done) => {
        let wavedata = {
            "enemy_amount": 0,
            "time_between_enemies": 2.3,
            "difficulty": "extreme"
        }

        try {
            const response = await request.post(`/createWave/game1`).send(wavedata)
            expect(response.status).toBe(201);
            done()
        } catch (error) {
            console.log(error);
        }
    })
    test('if wave exists', async (done) => {
        let data = {
            "enemy_amount": 0,
            "time_between_enemies": 2.3,
            "difficulty": "extreme"
        }

        try {
            const response = await request.post(`/createWave/game1`).send(data)
            expect(response.status).toBe(500);
            done()
        } catch (error) {
            console.log(error);
        }
    })


    test('if the game for the new wave exists', async (done) => {
        let data = {
            "enemy_amount": 5,
            "time_between_enemies": 10.2,
            "difficulty": "medium"
        }

        try {
            const response = await request.post(`/createWave/bestaatNiet`).send(data)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })


})


api/src/__tests__/end-to-end.test.js
const supertest = require("supertest");
const app = require("../server.js");
const request = supertest(app);


/**
* FLOW of tests
* ----------------------------------------------
* create game
* create wave on that specific game
* get all waves and check if that game is included
* update the wave
* delete the game
* check waves and games if all are deleted
*/



describe('/creategame/', () => {
    let gamedata = {
        title: "little big planet",
        summary: "coop adventure game where u play as a customizable character"
    }
    let waveuuid;
    let gameuuid;

    //create game
    test('create game', async (done) => {
        try {
            const response = await request.post(`/createGame`).send(gamedata)
            expect(response.status).toBe(201);
            gameuuid = (response.body.uuid);
            done()
        } catch (error) {
            console.log(error)
        }
    })

    //create wave 
    let wavedata = {
        "enemy_amount": 12,
        "time_between_enemies": 1.2,
        "difficulty": "hard"
    }
    test('create wave on "little big planet"', async (done) => {

        try {
            const response = await request.post(`/createWave/little big planet`).send(wavedata)
            expect(response.status).toBe(201);

            done()
        } catch (error) {
            console.log(error);
        }
    })
    //get waves of little big planet
    test('if little big planet exists and wave was added ', async (done) => {
        try {
            const response = await request.get('/getWavesFromGame/little big planet')

            waveuuid = response.body[0].uuid;
            expect(response.status).toBe(200);
            done()
        } catch (error) {

        }
    })

    //update the wave
    test('if wavedifficulty is updated', async (done) => {
        let data = {
            "difficulty": "extreme"
        }
        try {

            const response = await request.patch(`/changeWave/${waveuuid}`).send(data)
            expect(response.status).toBe(204);
            done()
        } catch (error) {
            console.log(error);
        }
    })
    //delete game
    test('if game is deleted', async (done) => {

        try {

            const response = await request.delete(`/deleteGame/${gameuuid}`)
            expect(response.status).toBe(200);
            done()
        } catch (error) {
            console.log(error);
        }
    })

    //get waves of little big planet
    test('if little big planet doesnt exist anymore and waves were removed ', async (done) => {
        try {
            const response = await request.get('/getWavesFromGame/little big planet')


            expect(response.status).toBe(404);
            done()
        } catch (error) {

        }
    })


})

api/src/__tests__/integration.test.js
const supertest = require("supertest");
const app = require("../server.js");
const request = supertest(app);


///DELETE ENTIRE GAME///
describe('/createWave/:gameTitle', () => {
    let data_1 = {
        "enemy_amount": 5,
        "time_between_enemies": 10.2,
        "difficulty": "medium"
    }

    test('return 500 when game doesnt exist', async (done) => {
        try {
            const response = await request.post(`/createWave/bestaatNiet`).send(data_1)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })

    let data_2 = {
        "enemy_amount": 5,
        "time_between_enemies": 10.2,
        "difficulty": "madium"
    }
    test('return 500 when difficulty is other than easy/medium/hard/extreme', async (done) => {
        try {
            const response = await request.post(`/createWave/call of duty`).send(data_2)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })
    let data_3 = {
        "enemy_amount": -5,
        "time_between_enemies": 10.2,
        "difficulty": "medium"
    }
    test('return 500 when enemy_amount is smaller than 0', async (done) => {
        try {
            const response = await request.post(`/createWave/call of duty`).send(data_3)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })
    let data_4 = {
        "enemy_amount": "invalid",
        "time_between_enemies": 10.2,
        "difficulty": "medium"
    }
    test('return 500 when enemy_amount is no number', async (done) => {
        try {
            const response = await request.post(`/createWave/call of duty`).send(data_4)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })
    let data_5 = {
        "enemy_amount": 5,
        "time_between_enemies": -3,
        "difficulty": "medium"
    }
    test('return 500 when time_between_enemy is smaller than 0', async (done) => {
        try {
            const response = await request.post(`/createWave/call of duty`).send(data_5)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })
    let data_6 = {
        "enemy_amount": 6,
        "time_between_enemies": "invalid",
        "difficulty": "medium"
    }
    test('return 500 when time_between_enemy is no number', async (done) => {
        try {
            const response = await request.post(`/createWave/call of duty`).send(data_6)
            expect(response.status).toBe(500);
            done()

        } catch (error) {
            console.log(error);
        }
    })
})

api/src/server.js
const express = require('express');
const bodyParser = require('body-parser');
const http = require('http');
const uuidHelper = require("./utils/UuidHelper.js");
const { debug } = require('console');
const { randomInt } = require('crypto');
const cors = require('cors')



const pg = require("knex")({
    client: 'pg',
    version: '9.6',
    searchPath: ['knex', 'public'],
    connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/easywavespawnerdb'
});



const app = express();
http.Server(app);
app.use(cors())


app.use(bodyParser.json());
app.use(
    bodyParser.urlencoded({
        // to support URL-encoded bodies
        extended: true
    })
);



app.get('/', (req, res) => {
    res.sendStatus(200);
})

///GET ALL WAVES///
/**
    * @param none
    * @returns array of objects
*/
app.get('/getAllWaves', async (req, res) => {
    try {
        const result = await pg.select(["uuid", "game_id", "enemy_amount", "difficulty", "time_between_enemies"]).from("waves");

        res.json(
            result
        );
    } catch (error) {
        res.send(error)
    }

})
///GET ALL GAMES///
/**
    * @param none
    * @returns array of objects
*/
app.get('/getAllGames', async (req, res) => {
    try {
        const result = await pg.from("games");

        res.json({
            res: result,
        });
    } catch (error) {
        res.send(error)
    }

})
///GET ALL WAVES FROM SPECIFIC GAME///
/**
    * @param title => gametitle
    * @returns array of objects => empty if no game found
*/
app.get('/getWavesFromGame/:title', async (req, res) => {
    let titletocheck = req.params.title.toLowerCase();
    try {
        const result = await pg
            .select(["waves.id", "waves.uuid", "waves.enemy_amount", "waves.difficulty", "waves.time_between_enemies"])
            .from("games")
            .rightJoin('waves', 'waves.game_id', 'games.id').where({ title: titletocheck });
        if (result.length == 0) {
            res.sendStatus(404);
        } else {
            res.json(
                result
            )
        }
    } catch (error) {
        res.send(error)
    }
})
///GET ALL WAVES BY DIFFICULTY///
/**
    * @param difficulty => easy/medium/hard/extreme
    * @returns array of objects => empty if no wave found
*/
app.get('/getWavesByDifficulty/:difficulty', async (req, res) => {
    let stringToCheck = req.params.difficulty.toLowerCase();
    if (stringToCheck !== "easy" && stringToCheck !== "medium" && stringToCheck !== "hard" && stringToCheck !== "extreme") {
        res.send(false)
    } else {

        try {
            const result = await pg
                .from("waves")
                .where({ difficulty: stringToCheck })
            res.json({
                waves: result
            })
        } catch (error) {
            res.send(error)
        }
    }
})
/////DELETE/////
///DELETE WAVE BY ID///
/**
    * @param id => game uuid
    * @returns if succeeded => id of game | if not succeeded => 404 | if wave doesn't exist =>"wave doesn't exist"
*/
app.delete('/deleteWaveByid/:id', async (req, res) => {
    try {
        const result = await pg.from("waves").where({ uuid: req.params.id })
        if (result.length !== 0) {

            await pg
                .table("waves")
                .where({ uuid: req.params.id })
                .del().then(() => {
                    console.log(`deleted ${req.params.id} from waves`);
                    res.send(`deleted ${req.params.id} from waves`);
                }).catch((e) => {
                    console.log(e);
                })
        } else {
            res.send("wave doesn't exist");
        }


    } catch (error) {
        res.send(error)
    }
})
///DELETE ENTIRE GAME///
/**
    * @param id => uuid of game
    * @returns if succeeded => id of game | if not succeeded => 404 | if game doesn't exist =>"game doesn't exist"
*/
app.delete('/deleteGame/:id', async (req, res) => {
    try {
        const result = await pg.from("games").where({ uuid: req.params.id })
        if (result.length !== 0) {

            await pg
                .table("games")
                .where({ uuid: req.params.id })
                .del().then(() => {
                    console.log(`deleted ${req.params.id} from games`);
                    res.send(`deleted ${req.params.id} from games`);
                }).catch((e) => {
                    console.log(e);
                })
        } else {
            res.send("wave doesn't exist");
        }


    } catch (error) {
        res.send(error)
    }
})
/////UPDATE/////
///ADJUST WAVE DIFFICULTY///
/**
    * @param id => uuid of wave | body = object with change params {difficulty:string,enemy_amount:number,time_between_enemies:number}
    * @returns if params are bad => false, succeeded => 204, not found => "wave doesnt exist"
*/
app.patch('/changeWave/:id', async (req, res) => {
    let canpass = true;
    if (req.body.hasOwnProperty("difficulty")) {
        let newdifficulty = req.body.difficulty.toLowerCase();
        if (newdifficulty !== "easy" && newdifficulty !== "medium" && newdifficulty !== "hard" && newdifficulty !== "extreme") {
            canpass = false

        } else {
            req.body.difficulty = newdifficulty
        }

    }
    if (req.body.hasOwnProperty("enemy_amount")) {
        let newEnemyAmount = req.body.enemy_amount
        if (typeof newEnemyAmount !== "number") {
            canpass = false
        }
    }
    if (req.body.hasOwnProperty("time_between_enemies")) {
        let newTime = req.body.time_between_enemies
        if (typeof newTime !== "number") {
            canpass = false
        }

    }

    if (canpass === false) {
        res.send(canpass);
    } else {

        try {
            const result = await pg.from("waves").where({ uuid: req.params.id })
            if (result.length !== 0) {
                await pg
                    .table("waves")
                    .where({ uuid: req.params.id })
                    .update(req.body).then(() => {
                        res.send(204);
                    }).catch((e) => {
                        console.log(e);
                    })
            } else {
                res.send("wave doesn't exist");
            }


        } catch (error) {
            res.send(error)
        }
    }
})
/////CREATE/////
///CREATE NEW GAME///
/**
    * @param body (object) => {title:string,summary:string}
    * @returns 201 (succeeded), 500 (if game already exists),500 (if game title is "")
*/
app.post('/createGame', async (req, res) => {
    //no caps
    req.body.title = req.body.title.toLowerCase();
    if (req.body.title !== "") {

        try {
            const result = await pg.from("games").where({ title: req.body.title });
            console.log(result.length);
            if (result.length === 0) {

                try {
                    const uuid = uuidHelper.generateUUID();
                    await pg.table("games").insert({ uuid, title: req.body.title, summary: req.body.summary }).then(() => {

                        res.status(201).send({ uuid: uuid });
                    })

                } catch (error) {
                    res.send(error)
                }
            } else {

                res.status(500).send();
            }
        } catch (error) {
            res.send(error);
        }
    } else {
        res.status(500).send("title of game needs to be at least 1 character long");
    }
    //check if already exist
})
///CREATE NEW WAVE///
/**
    * @param gameTitle => name of game | body = object with params {difficulty:string,enemy_amount:number,time_between_enemies:number}
    * @returns 201 (succeeded), 500 (if wave already exists),500 (if game title is not found)
*/
app.post('/createWave/:gameTitle', async (req, res) => {
    //no caps in difficulty
    req.body.difficulty = req.body.difficulty.toLowerCase();
    req.params.gameTitle = req.params.gameTitle.toLowerCase();
    let gameID;
    let canpass = true;
    console.log(req.body);
    if (req.body.hasOwnProperty("difficulty")) {
        let newdifficulty = req.body.difficulty.toLowerCase();
        if (newdifficulty !== "easy" && newdifficulty !== "medium" && newdifficulty !== "hard" && newdifficulty !== "extreme") {
            canpass = false


        } else {
            req.body.difficulty = newdifficulty
        }

    }
    if (req.body.hasOwnProperty("enemy_amount")) {
        let newEnemyAmount = req.body.enemy_amount
        if (typeof newEnemyAmount !== "number") {
            canpass = false

        }
        if (newEnemyAmount < 0) {
            canpass = false

        }

    }
    if (req.body.hasOwnProperty("time_between_enemies")) {
        let newTime = req.body.time_between_enemies
        if (typeof newTime !== "number") {
            canpass = false

        }
        if (newTime < 0) {
            canpass = false

        }

    }
    if (canpass) {


        //if game exist => make wave with id of game
        try {
            let result = await pg.select(["id"]).from("games").where({ title: req.params.gameTitle });



            if (result.length !== 0) {
                gameID = result[0].id;

                try {
                    const allwaves = await pg.from("waves").where({ game_id: gameID, difficulty: req.body.difficulty, enemy_amount: req.body.enemy_amount, time_between_enemies: req.body.time_between_enemies });
                    console.log(allwaves.length);
                    if (allwaves.length === 0) {
                        try {
                            const uuid = uuidHelper.generateUUID();
                            await pg.table("waves").insert({ game_id: gameID, uuid, time_between_enemies: req.body.time_between_enemies, enemy_amount: req.body.enemy_amount, difficulty: req.body.difficulty }).then(() => {
                                res.status(201).send();
                            })

                        } catch (error) {
                            res.send(error)
                        }
                    }
                    else {
                        res.status(500).send("wave already exist")
                    }
                } catch (error) {

                }
            } else {
                res.status(500).send("game not found");
            }
        } catch (error) {

        }
    } else { res.status(500).send() }

})

///INITIALISETABLES IF THEY DON'T EXIST///
async function initialiseTables() {

    await pg.schema.hasTable('games').then(async (exists) => {
        if (!exists) {
            try {
                await pg.schema
                    .createTable('games', (table) => {
                        table.increments("id")
                        table.uuid('uuid');
                        table.string('title');
                        table.string('summary');
                    })
                    .then(async () => {
                        for (let index = 0; index < 2; index++) {
                            const uuid = uuidHelper.generateUUID();
                            await pg.table('games').insert({ uuid, title: `game${index + 1}`, summary: `some summary of game${index + 1}` });
                        }
                    });
            } catch (error) {
                console.log(error);
            }

        }
    });
    await pg.schema.hasTable('waves').then(async (exists) => {
        if (!exists) {
            try {
                await pg.schema
                    .createTable('waves', (table) => {
                        table.increments();
                        table.uuid('uuid');
                        table.integer("game_id").unsigned().references("id").inTable("games").onDelete("CASCADE").onUpdate("CASCADE");
                        table.integer('enemy_amount');
                        table.string('difficulty');
                        table.float('time_between_enemies');
                    }).then(async () => {

                        const result = await pg
                            .select(['id'])
                            .from('games');
                        console.log(result);
                        for (let index = 0; index < 10; index++) {
                            const uuid = uuidHelper.generateUUID();
                            await pg.table('waves').insert({
                                uuid, game_id: result[Math.floor(Math.random() * 2)].id, enemy_amount: 0, difficulty: "hard", time_between_enemies: 2.5
                            });
                        }
                        process.exit();

                    });

            } catch (error) {
                console.log(error);
            }

        }
    });



}
initialiseTables()



module.exports = app;